### ISSUE TITLE
Transmission 3.0.0 causes an exception if you want to get torrent/tracker details

### ISSUE THREAD

**Issue by gregorkistler** [openseedbox/openseedbox#71]
As mentioned by @stickz in #69 Transmission 3.00 breaks the popup window for torrent/tracker details. If you click at the ℹ️  button you end up getting a notification at the top telling `Unsuccessful webservice call, status:` as you can see here:

<img width="974" alt="Screenshot_2021-01-02_at_13_12_20" src="https://user-images.githubusercontent.com/12155030/103457283-2882d880-4cfe-11eb-8e93-396c177f7c93.png">

The corresponding .log file contains an exception:

```
Jan  2 13:11:37 172.21.0.1 backend[6682]: 12:11:37,751 ERROR ~ 
Jan  2 13:11:37 172.21.0.1 backend[6682]: 
Jan  2 13:11:37 172.21.0.1 backend[6682]: @7iabggodg
Jan  2 13:11:37 172.21.0.1 backend[6682]: Internal Server Error (500)
Jan  2 13:11:37 172.21.0.1 backend[6682]: 
Jan  2 13:11:37 172.21.0.1 backend[6682]: Execution exception (In /app/com/openseedbox/backend/transmission/TransmissionBackend.java around line 426)
Jan  2 13:11:37 172.21.0.1 backend[6682]: JsonSyntaxException occurred : java.lang.IllegalStateException: Expected NUMBER but was BOOLEAN at path $.trackerStats[0].lastScrapeTimedOut
Jan  2 13:11:37 172.21.0.1 backend[6682]: 
Jan  2 13:11:37 172.21.0.1 backend[6682]: play.exceptions.JavaExecutionException: java.lang.IllegalStateException: Expected NUMBER but was BOOLEAN at path $.trackerStats[0].lastScrapeTimedOut
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invoke(ActionInvoker.java:485)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:467)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:436)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invoke(ActionInvoker.java:159)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.server.ServletWrapper$ServletInvocation.execute(ServletWrapper.java:579)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.Invoker$Invocation$1.apply(Invoker.java:302)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.Invoker$Invocation$1.apply(Invoker.java:298)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.db.jpa.JPA.withTransaction(JPA.java:316)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.db.jpa.JPA.withinFilter(JPA.java:212)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.db.jpa.JPAPlugin$TransactionalFilter.withinFilter(JPAPlugin.java:292)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.Invoker$Invocation.withinFilter(Invoker.java:281)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.Invoker$Invocation.run(Invoker.java:298)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.server.ServletWrapper$ServletInvocation.run(ServletWrapper.java:570)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.Invoker.invokeInThread(Invoker.java:70)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.server.ServletWrapper.service(ServletWrapper.java:157)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:888)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1597)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at java.lang.Thread.run(Thread.java:748)
Jan  2 13:11:37 172.21.0.1 backend[6682]: Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected NUMBER but was BOOLEAN at path $.trackerStats[0].lastScrapeTimedOut
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.Gson.fromJson(Gson.java:887)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.Gson.fromJson(Gson.java:952)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.Gson.fromJson(Gson.java:925)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.openseedbox.backend.transmission.TransmissionBackend.getTorrentInfo(TransmissionBackend.java:426)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.openseedbox.backend.transmission.TransmissionBackend.getTorrentTrackers(TransmissionBackend.java:244)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at controllers.Torrents.trackers(Torrents.java:164)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invokeWithContinuation(ActionInvoker.java:536)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at play.mvc.ActionInvoker.invoke(ActionInvoker.java:473)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011... 37 more
Jan  2 13:11:37 172.21.0.1 backend[6682]: Caused by: java.lang.IllegalStateException: Expected NUMBER but was BOOLEAN at path $.trackerStats[0].lastScrapeTimedOut
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.JsonTreeReader.nextInt(JsonTreeReader.java:241)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.TypeAdapters$7.read(TypeAdapters.java:243)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.TypeAdapters$7.read(TypeAdapters.java:235)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
Jan  2 13:11:37 172.21.0.1 backend[6682]: #011... 50 more
```

Similar issues were also reported in other projects e.g. via https://github.com/metalmatze/transmission-exporter/issues/26

Main cause:
https://github.com/transmission/transmission/commit/10cdd7f790eed859ad333a6a7635a2d6a2d53447#diff-f5f6943a983d494a06191cbf0086b2ddd165e6e846dc085d5833abd63683cc3dL492

Fix:
By changing the type from `int` to `boolean` in https://github.com/openseedbox/openseedbox-server/blob/0853ae1583436771b501ac987eccac0e4399e5cc/app/com/openseedbox/backend/transmission/TransmissionTracker.java#L24 and recompiling the application it does work as expected.

**Comment by gregorkistler** [comment]
As we cannot force users to use the latest (at time of writing) transmission-daemon version we need to check the installed package version.

**Comment by xabolcs** [comment]
Thanks for the report!

**Comment by xabolcs** [comment]
> As we cannot force users ...

While fixing it to keep compatibility with Transmission < 3.0.0, it would be nice to create tests with old and new payloads: one commit for tests, one commit for fix

**Comment by xabolcs** [comment]
gson [User Guide](https://github.com/google/gson/blob/master/UserGuide.md).

**Comment by xabolcs** [comment]
Payloads for tests

`transmission-daemon 2.94 (d8e60ee44f)`:
```
11:51:39,016 INFO  ~ Sending request: {"method":"torrent-get","arguments":{"ids":["65162b3234bbafbc71beb372562f59148a1526be"],"fields":["hashString","trackerStats"]}}
11:51:39,018 INFO  ~ Response: [{"hashString":"65162b3234bbafbc71beb372562f59148a1526be","trackerStats":[{"announce":"http://bttracker.debian.org:6969/announce","announceState":1,"downloadCount":141,"hasAnnounced":true,"hasScraped":true,"host":"http://bttracker.debian.org:6969","id":0,"isBackup":false,"lastAnnouncePeerCount":32,"lastAnnounceResult":"Success","lastAnnounceStartTime":1609761014,"lastAnnounceSucceeded":true,"lastAnnounceTime":1609761015,"lastAnnounceTimedOut":false,"lastScrapeResult":"","lastScrapeStartTime":1609761020,"lastScrapeSucceeded":true,"lastScrapeTime":1609761021,"lastScrapeTimedOut":0,"leecherCount":1,"nextAnnounceTime":1609761915,"nextScrapeTime":1609762830,"scrape":"http://bttracker.debian.org:6969/scrape","scrapeState":1,"seederCount":44,"tier":0}]}]
```


`transmission-daemon 3.00 (bb6b5a062e)`:
```
12:09:34,997 INFO  ~ Sending request: {"method":"torrent-get","arguments":{"ids":["65162b3234bbafbc71beb372562f59148a1526be"],"fields":["hashString","trackerStats"]}}
12:09:34,998 INFO  ~ Response: [{"hashString":"65162b3234bbafbc71beb372562f59148a1526be","trackerStats":[{"announce":"http://bttracker.debian.org:6969/announce","announceState":0,"downloadCount":-1,"hasAnnounced":false,"hasScraped":false,"host":"http://bttracker.debian.org:6969","id":0,"isBackup":false,"lastAnnouncePeerCount":0,"lastAnnounceResult":"","lastAnnounceStartTime":0,"lastAnnounceSucceeded":false,"lastAnnounceTime":0,"lastAnnounceTimedOut":false,"lastScrapeResult":"","lastScrapeStartTime":0,"lastScrapeSucceeded":false,"lastScrapeTime":0,"lastScrapeTimedOut":false,"leecherCount":-1,"nextAnnounceTime":0,"nextScrapeTime":1609762250,"scrape":"http://bttracker.debian.org:6969/scrape","scrapeState":1,"seederCount":-1,"tier":0}]}]
```

**Comment by xabolcs** [comment]
Fix is ready,
```
$ git diff --stat app/com/openseedbox/backend/
 app/com/openseedbox/backend/transmission/TransmissionBackend.java | 32 ++++++++++++++++++++++++++++----
 app/com/openseedbox/backend/transmission/TransmissionTracker.java |  2 +-
 2 files changed, 29 insertions(+), 5 deletions(-)
```
.. but I'm unable to create a test which drives the bugging `getTorrentInfo`. :(

Some SO answers recommends [Rest Driver](https://github.com/rest-driver/rest-driver/wiki/Client-driver). Any other hint?

**Comment by xabolcs** [comment]
> Any other hint?

After reading a lot, I'm going the [`Router.addRoute()`](https://github.com/playframework/play1/blob/f818294c6786ce5d490e8bb70ef54d7091d96b40/modules/testrunner/src/play/modules/testrunner/TestRunnerPlugin.java#L25-L32) way, checking [`Play.id`](https://github.com/playframework/play1/blob/f818294c6786ce5d490e8bb70ef54d7091d96b40/framework/src/play/Play.java#L82-L85) if it's a [isTestFrameworkId](https://github.com/playframework/play1/blob/f818294c6786ce5d490e8bb70ef54d7091d96b40/framework/pym/play/utils.py#L241-L242).

Of course, a `TestTransmissionBackendController.java` and a [`%test.backend.transmission.port=....`](https://github.com/openseedbox/openseedbox-server/blob/cc60e57190a4d1b96843d78d818520d77097f7ce/conf/application.conf.default#L29-L30) in `application.conf` with a few routes needs to be done.

**Comment by xabolcs** [comment]
:roll_eyes: 

It works for manual requests, but it times out / responses null through TestRunner (http://localhost:9001/@tests).
I don't know what's wrong.


Pushed to my [branch](https://github.com/openseedbox/openseedbox-server/compare/master...xabolcs:branch-issue-issue-71-transmission-3-rpc-int-to-boolean)!

**Comment by xabolcs** [comment]
> Some SO answers recommends [Rest Driver](https://github.com/rest-driver/rest-driver/wiki/Client-driver). Any other hint?

There is WireMock too: https://wiremock.org / https://github.com/tomakehurst/wiremock

**Comment by xabolcs** [comment]
I will polish the fix' commit message and push it without creating a test. :roll_eyes: 

For the test, I'm going to create another issue.

**Comment by gregorkistler** [comment]
Alright, thanks for taking care of 👍

**Comment by xabolcs** [comment]
Fixed by commit openseedbox/openseedbox-server@9c153f4157077b2e235b1f4695455de3fc7cf08f !

**Comment by stickz** [comment]
> As we cannot force users to use the latest (at time of writing)...

I tested transmission 3.00 on openseedbox for the past month. It's not stable. There's frequent crashes. I'm going back to the precompiled stable 2.94 version from Debian Buster. I'll give it anther spin when Debian Bullseye is released.

> Fixed by commit [openseedbox/openseedbox-server@9c153f4](https://github.com/openseedbox/openseedbox-server/commit/9c153f4157077b2e235b1f4695455de3fc7cf08f) !

Thanks for fixing this issue. This will be useful when Debian Bullseye is released, hopefully in the not too distant future.

**Comment by xabolcs** [comment]
Thank you for testing with Transmission 3.00!

**Comment by stickz** [comment]
I'm testing a possible fix for transmission 3.0 currently. I changed my docker build to `ubuntu:latest`, so I could use the ubuntu transmissionbt ppa. This source seems to be more stable for transmission-daemon, than my custom built version.

It's fully compatible with the current openseedbox configuration. Just need to add the ubuntu transmissionbt ppa in `openseedbox-server`. The image size is only slightly higher. Nginx runs faster because it's compiled with GCC 9.2 instead of 8.3.
